frontend/javascripts/components/async_clickables.js

```diff
@@ -1,74 +1,68 @@
 // @flow
-/* eslint-disable react/jsx-no-bind  */
-import { Button, Spin } from "antd";
+import { Button, Icon } from "antd";
 import * as React from "react";
-
-const onClick = async function(event: SyntheticInputEvent<>) {
-  this.setState({ isLoading: true });
-  try {
-    await this.props.onClick(event);
-  } finally {
-    if (this._isMounted) {
-      this.setState({ isLoading: false });
-    }
-  }
-};
+const { useState, useEffect, useRef } = React;
 
 type Props = {
   onClick: (SyntheticInputEvent<>) => Promise<any>,
 };
 
-type State = { isLoading: boolean };
-
-export class AsyncButton extends React.PureComponent<Props, State> {
-  _isMounted: boolean;
-  state = { isLoading: false };
+function useLoadingClickHandler(originalOnClick: (SyntheticInputEvent<>) => Promise<any>) {
+  const [isLoading, setIsLoading] = useState(false);
+  const wasUnmounted = useRef(false);
 
-  componentDidMount() {
-    this._isMounted = true;
-  }
-
-  componentWillUnmount() {
-    this._isMounted = false;
-  }
+  useEffect(
+    () => () => {
+      wasUnmounted.current = true;
+    },
+    [],
+  );
 
-  render() {
-    return <Button {...this.props} loading={this.state.isLoading} onClick={onClick.bind(this)} />;
-  }
-}
-
-export class AsyncLink extends React.PureComponent<Props & { children: React.Node }, State> {
-  _isMounted: boolean;
-  static defaultProps = {
-    children: [],
+  const onClick = async (event: SyntheticInputEvent<>) => {
+    setIsLoading(true);
+    try {
+      await originalOnClick(event);
+    } finally {
+      if (!wasUnmounted.current) {
+        setIsLoading(false);
+      }
+    }
   };
+  return [isLoading, onClick];
+}
 
-  state = { isLoading: false };
+export function AsyncButton(props: Props) {
+  const [isLoading, onClick] = useLoadingClickHandler(props.onClick);
+  return <Button {...props} loading={isLoading} onClick={onClick} />;
+}
 
-  componentDidMount() {
-    this._isMounted = true;
-  }
+export function AsyncIconButton(allProps: Props & { type: string }) {
+  const { type, ...props } = allProps;
+  const [isLoading, onClick] = useLoadingClickHandler(props.onClick);
+  return <Icon {...props} type={isLoading ? "loading" : type} onClick={onClick} />;
+}
 
-  componentWillUnmount() {
-    this._isMounted = false;
+export function AsyncLink(props: Props & { children: React.Node }) {
+  const [isLoading, onClick] = useLoadingClickHandler(props.onClick);
+  let content;
+  if (isLoading) {
+    const children = React.Children.toArray(props.children);
+    const childrenWithoutIcon = children.filter(child => {
```

Comment:
I'm not sure whether this actually works, at least in my testing for archiving and reopening explorative tracings, the loading icon appeared left of the icon that was there anyways. However, if you ask me that's totally fine, your call :)

Reply:
Ah, damnit. It doesn't work if the code is minified, since `child.type` is a function whose name gets lost during compilation. I'll leave it as is for now :see_no_evil: